File: leetcode/medium/39. Combination Sum/solution.cc
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> res;
    vector<int> temp;

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        pick(candidates, target, 0);
        return res;
    }

    void pick(vector<int>& candidates, int target, int index) {
        if (target == 0) {
            res.push_back(temp);
            return;
        }

        if (index == candidates.size() || target < 0) {
            return;
        }

        temp.push_back(candidates[index]);
        pick(candidates, target - candidates[index], index);
        temp.pop_back();
        pick(candidates, target, index + 1);
    }
};

File: leetcode/medium/39. Combination Sum/solution.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
function combinationSum(candidates, target) {
    const temp = [], res = [];
    let sum = 0;
    const pick = (index) => {
        if (sum >= target) {
            if (sum === target) res.push(temp.slice());
            return;
        }
        for (let i = index; i < candidates.length; i += 1) {
            temp.push(candidates[i])
            sum += candidates[i];
            pick(i);
            sum -= candidates[i];
            temp.pop();
        }
    }
    pick(0);

    return res;
};